volatility3/framework/symbols/linux/utilities/module_extract.py

```diff
@@ -0,0 +1,891 @@
+            )
+            return None
+
+        return data
+
+    @classmethod
+    def _make_section_header_64(
+        cls,
+        name_index: int,
+        name: str,
+        address: int,
+        size: int,
+        file_offset: int,
+        strtab_index: int,
+        symtab_index: int,
+    ) -> Optional[bytes]:
+        """
+        Creates a section header (Elf64_Shdr) for the given section
+        """
+        sect_header_type_int = cls._calc_sect_type(name)
+
+        flags = cls._calc_sect_flags(name)
+
+        link = cls._calc_link(name, strtab_index, symtab_index, sect_header_type_int)
+
+        entsize = cls._calc_entsize(name, sect_header_type_int, 64)
+
+        try:
+            sh_name = struct.pack("<I", name_index)
+            sh_type = struct.pack("<I", sect_header_type_int)
+            sh_flags = struct.pack("<Q", flags)
+            sh_addr = struct.pack("<Q", address)
+            sh_offset = struct.pack("<Q", file_offset)
+            sh_size = struct.pack("<Q", size)
+            sh_link = struct.pack("<I", link)
+            sh_info = b"\x00" * 4
+            sh_addralign = b"\x01\x00\x00\x00\x00\x00\x00\x00"
+            sh_entsize = struct.pack("<Q", entsize)
+
+        # catch overflows of offset/address/size
+        except struct.error:
+            vollog.debug(
+                f"Unable to build section header for section {name} at address {address:#x}"
+            )
+            return None
+
+        data = (
+            sh_name
+            + sh_type
+            + sh_flags
+            + sh_addr
+            + sh_offset
+            + sh_size
+            + sh_link
+            + sh_info
+            + sh_addralign
+            + sh_entsize
+        )
+
+        # This should never happen regardless of smear or other issues in the data. We build the structure to spec.
+        if len(data) != 64:
+            vollog.error(
+                f"Size of section data is {len(data)} expected 64 for section {name} at address {address:#x}"
+            )
+            return None
+
+        return data
+
+    @classmethod
+    def extract_module(
+        cls,
+        context: interfaces.context.ContextInterface,
+        vmlinux_name: str,
+        module: extensions.module,
+    ) -> Optional[bytes]:
+        # Bail early if bad address sent in
+        try:
+            hasattr(module.sect_attrs, "nsections")
+        except exceptions.InvalidAddressException:
+            vollog.debug(f"module at offset {module.vol.offset:#x} is paged out.")
+            return None
+
+        # Gather sections
+        updated_sections, strtab_index, symtab_index = cls._parse_sections(
+            context, vmlinux_name, module
+        )
+
+        kernel = context.modules[vmlinux_name]
+
+        # Figure out header sizes
+        if symbols.symbol_table_is_64bit(context, kernel.symbol_table_name):
+            make_elf_header = cls._make_elf_header_64
```

Comment:
This is nice, still feels like it could be a single param handed to a single function though.   Happy to go with whatever you decide, just want a written rationale for why one way over the other...

Reply:
This way so the specific 32/64 headers are made correctly and in line with their spec in a direct flow

Reply:
I don't see how it wouldn't be a direct flow if they both went into the same function and then the differences were decided within that one function?  Feels like it would clarify for people looking into it how the two differ and possibly why.  I'll leave it to you, but it just seems to increase the code and spread it out, rather than having it very clearly share all but the parts that differ (which would then be clearly visible by `x if 32bit else y` or similar...